mephisto/core/task_launcher.py

```diff
@@ -140,28 +174,33 @@ def generate_units(self):
                     yield unit
                 else:
                     break
-            for db_id in units_id_to_remove:
-                self.unlaunched_units.pop(db_id)
+            with self.unlaunched_units_access_condition:
+                for db_id in units_id_to_remove:
+                    self.unlaunched_units.pop(db_id)
 
             time.sleep(UNIT_GENERATOR_WAIT_SECONDS)
             if not self.unlaunched_units:
                 break
 
-    def _try_launch_units(self, url: str) -> None:
-        """ Launch all units or limited number of units according to (max_num_concurrent_units)"""
-        while self.assignment_exist:
-            if self.max_num_concurrent_units > 0:
-                for unit in self.generate_units():
-                    unit.launch(url)
-            else:
-                for _ in range(len(self.unlaunched_units.items())):
-                    db_id, unit = self.unlaunched_units.popitem()
-                    unit.launch(url)
-                time.sleep(0.5)
+    def _launch_limited_units(self, url: str) -> None:
+        """ use units' generator to launch limited number of units according to (max_num_concurrent_units)"""
+        while self.did_launch_units:
```

Comment:
What is this flag actually doing?

Reply:
it was somehow making sure the units generator keeps working .. but now I realized it is not needed at all. Thanks!